### ISSUE TITLE
Only symlink the folder instead of hardlinking every single file

### ISSUE THREAD

**Issue by aminya** [pnpm/pnpm#3109]
I have noticed that pnpm hardlinks every single separate file separately. This results in lower performance when the package is `require`d because the `realpath` function should be called every time for each single file.

I am proposing to only symlink the top-level folder to fix this issue. 

### Expected/Actual behavior:

These should be real files, not hardlinks. Only the top-level folder should be symlinked:
![image](https://user-images.githubusercontent.com/16418197/106150083-95905f00-6140-11eb-9b0f-e2c5fe5c66ea.png)

### Additional information:

 - `node -v` prints: 15
 - Windows, macOS, or Linux?: win 10

An example of lower performance is in Parcel builds [is this](https://github.com/parcel-bundler/parcel/issues/5072#issuecomment-761209246
). Using yarn results in ~3s faster build times.

Build time:
<table>
<tr>
	<td> pnpm
	<td> yarn
<tr>
	<td> 10.5 s
	<td> 7.4 s
</table>

**Comment by zkochan** [comment]
I don't understand this. You probably mean symlinks, not hard links. It is not possible to create a hard link of a directory. And pnpm only creates symlinks for directories, so I don't understand how you got symlinked files as on the screenshot.

**Comment by aminya** [comment]
Yes. I was loose with my wording. I am proposing to symlink the top-level folder to the actual files. Currently, the folders are symlinked and inside the folder, there are hardlinks again.

**Comment by zkochan** [comment]
But hard links are real files. They are just pointing to the same location on the disk. I don't understand why your file explorer marks those files with some arrows. To any program, a hard link is just a file. A regular file.

**Comment by aminya** [comment]
Maybe, I am not explaining myself correctly. My point is that there should be only "one indirection", and that is the folder symlink. Currently, even inside the symlinked folder there are new hardlinks. 

When I symlink a folder, this is how it looks:
![image](https://user-images.githubusercontent.com/16418197/106216086-f4cd8e00-6197-11eb-8653-e68336632829.png)
![image](https://user-images.githubusercontent.com/16418197/106216084-f303ca80-6197-11eb-9f43-764aabf805e1.png)

Just like the real file:
![image](https://user-images.githubusercontent.com/16418197/106216105-0151e680-6198-11eb-89dd-b8edf6345445.png)

**Comment by zkochan** [comment]
Here's a symlinked directory created by pnpm on my Windows machine, looks the same:

![image](https://user-images.githubusercontent.com/1927579/106217906-e9ea1c00-61de-11eb-8705-485939a43fb8.png)

I don't understand how you got something that looks like this:
![image](https://user-images.githubusercontent.com/1927579/106217942-fa9a9200-61de-11eb-9bc4-764435469c00.png)

**Comment by zkochan** [comment]
Symlink above, real location below:

![image](https://user-images.githubusercontent.com/1927579/106218047-35042f00-61df-11eb-8b6a-68ff1450771c.png)

**Comment by aminya** [comment]
I just use `pnpm i`. Here is an example:
![image](https://user-images.githubusercontent.com/16418197/106218277-a40c6400-619c-11eb-93a3-c4d6b0d71268.png)
package.json inside the folder is hardlinked:
![image](https://user-images.githubusercontent.com/16418197/106218351-c1d9c900-619c-11eb-815c-357eb4ce1a7d.png)

**Comment by zkochan** [comment]
ok, I see, your explorer UI marks hardlinks. OK. This is correct.

But this is not an issue because there *is* "one indirection". Node.js will resolve the symlinked directory to its real location and will resolve the file. Node.js is not looking for other locations of hard link.

**Comment by aminya** [comment]
Not sure about Node's `fs.readlpath.native`, but at least the function that Parcel uses has overhead. @mischnic might know better.

**Comment by aminya** [comment]
Since pnpm has bumped up the version, I did another benchmark of Parcel builds for [solid-simple-table](https://github.com/aminya/solid-simple-table). Using a yarn bootstrapped repository results in ~4s faster build times compared to pnpm bootstrapped repo.

Build time:
<table>
<tr>
	<td> pnpm
	<td> yarn
<tr>
	<td> 13.6 s
	<td> 9.18 s
</table>

**Comment by DanielRios549** [comment]
I think this should be the default behavior, the folder `node_modules/.pnpm` should not exists, or if exists, all the folder inside it should be symlinks do the original folders on the global store at `~/.pnpm-store`, this global store is where all the files should be store.

**Comment by jedwards1211** [comment]
@DanielRios549 doing it that way is not possible, it would cause problems with deduping module instances.

For example, in a React project, all components need to use the same instance of the `react` package, otherwise you'll get errors.

Let's say you have two projects:

### foo
```js
  "react": "16.8.0",
  "my-react-comp": "1.0.0",
```

### bar
```js
  "react": "17.0.0",
  "my-react-comp": "1.0.0"
```

With the behavior you want, `foo/node_modules/my-react-comp` and `bar/node_modules/my-react-comp` would have to be symlinks to the same directory within `~/.pnpm-store`, hence they would both be using the same version of `react`.  Assume that version is `16.8.0`.  Then the `foo` project would work fine, but in the `bar` project, `my-react-comp` would be loading `react` `16.8.0` -- different from the project's own version `17.0.0` that it uses elsewhere.

The only way to avoid this is for `pnpm` to hardlink the files for `my-react-comp` to the store rather than the entire directory, so that the files can be shared between both projects, but `my-react-comp` can require a different version of `react` in `foo` than it does in `bar`.

**Comment by Clashsoft** [comment]
The current behaviour is actually problematic with Angular, in particular ngcc, which writes new files into the packages. These new files don't end up in the global store and thus need to be generated anew all the time, especially on CI.

**Comment by everflux** [comment]
Avoiding symlinks would allow to use pnpm in a scenario with a container based build and mounting a global cache to the build container. 
symlinks can cross devices while hard links can not.
